ring/lifecycler.go

```diff
@@ -593,11 +593,46 @@ func (i *Lifecycler) initRing(ctx context.Context) error {
 			return ringDesc, true, nil
 		}
 
-		// If the ingester failed to clean its ring entry up in can leave its state in LEAVING
+		// If the ingester failed to clean its ring entry up it can leave its state in LEAVING
 		// OR unregister_on_shutdown=false
 		// Move it into ACTIVE to ensure the ingester joins the ring.
-		if instanceDesc.State == LEAVING && len(instanceDesc.Tokens) == i.cfg.NumTokens {
-			instanceDesc.State = ACTIVE
+		if instanceDesc.State == LEAVING {
+			var tokens Tokens = instanceDesc.Tokens // way of forcing tokens to be of type Tokens instead of []uint32.
+			setIsActive := true
+			if len(instanceDesc.Tokens) != i.cfg.NumTokens {
+				level.Debug(i.logger).Log("msg", "existing entry has different number of tokens", "existingTokens", len(instanceDesc.Tokens), "newTokens", i.cfg.NumTokens)
+				if len(tokensFromFile) > 0 {
+					level.Debug(i.logger).Log("msg", "adding tokens from file", "tokens", len(tokensFromFile))
+					setIsActive = len(tokensFromFile) >= i.cfg.NumTokens
+					tokens = tokensFromFile
+				} else if i.cfg.NumTokens > len(instanceDesc.Tokens) {
+					needTokens := i.cfg.NumTokens - len(instanceDesc.Tokens)
+					level.Debug(i.logger).Log("msg", "no tokens in file, generating new ones in addition to those of existing instance", "newTokens", needTokens)
+					newTokens := GenerateTokens(needTokens, ringDesc.GetTokens())
+					tokens = append(instanceDesc.Tokens, newTokens...)
+					sort.Sort(tokens)
+				} else {
+					level.Debug(i.logger).Log("msg", "no tokens in file, adopting a subset of existing instance's tokens", "numTokens", i.cfg.NumTokens)
+					tokens = instanceDesc.Tokens[0:i.cfg.NumTokens]
```

Comment:
Are the tokens sorted? If so, I think that this selects the smallest N, basically increasing the change that only smaller hashes will be selected for this instance, and bigger ones never match. To still have an even chance of distribution of hashes associated to all ingesters, I think random tokens should be left out of the slice, so the tokens should first be reshuffled, then take the first `i.cfg.NumTokens` and before assigning them to the instance sort them again.

Reply:
Good point @agardiman. I've implemented what you suggest, i.e. shuffle before dropping tokens. PTAL.